File: SAUCE/c++/skill.cpp
#include "skill.h"
#include <iostream>
Skill::Skill(double v, QString n,unsigned int m, double an, unsigned int cd,unsigned int r, double ps,unsigned int lv, bool re):MDamage(v), ready(re),Sname(n), manaCost(m), cooldown(cd), range(r), level(lv),projectileSpeed(ps),animation(an){
}


bool Skill::IsCastable(double time, unsigned int distance){
    if(time<animation+(distance/projectileSpeed) || distance>range || (level<=0 || level>4))
        return false;
    else
        return true;
}

bool Skill::IsCastable(double time, unsigned int distance,unsigned int mana){
    if(time<animation+(distance/projectileSpeed) || distance>range || (level<=0 || level>4)||mana<manaCost)
        return false;
    else
        return true;
}

double Skill::DPS(unsigned int distance){
        return DamageByTime(1,distance);
}

double Skill::DamageByTime(double time,unsigned int distance){
  return HitByTime(time,distance)*value;
}

unsigned int Skill::HitByTime(double time,unsigned int distance){
    double aux = time;
    unsigned int tiri = 0;
    while(IsCastable(aux,distance)){
        tiri++;
        aux=aux-(animation+cooldown+(distance/projectileSpeed));
        }
    return tiri;
}

double Skill::totalTime(unsigned int distance) const{
 return (distance/projectileSpeed)+animation;
}

bool Skill::Heal(){
    if(this->getValue()<0)
        return true;
    else
        return false;
}

double Skill::getAnimation() const {return animation;}
bool Skill::getReady() const {return ready;}
unsigned int Skill::getCooldown() const {return cooldown;}
unsigned int Skill::getManaCost() const {return manaCost;}
QString Skill::getName() const {return Sname;}
void Skill::setReady(bool b) {ready=b;}

double Skill::operator+(Damage* d) const{
    return this->getValue()+d->getValue();
}

double Skill::operator-(Damage* d) const{
     return this->getValue()-d->getValue();
}
bool Skill::operator==(Damage* d) const{
    Skill* x;
    x=dynamic_cast<Skill*>(d);
    if(x)
        return this->Sname==x->Sname;
    else
        return false;
}

bool Skill::operator==(Skill d) const{
    return this->Sname==d.Sname;
}
void Skill::editedValues(double nv,unsigned int nm,unsigned int na,unsigned int nc,unsigned int nr,double np,unsigned int nl){
    this->value = nv;
    this->manaCost = nm;
    this->animation = na;
    this->cooldown = nc;
    this->range = nr;
    this->projectileSpeed = np;
    this->level = nl;
}

Damage* Skill::sum(Damage * s) const{
    Skill* sum = new Skill;
    sum->value = *this + s ;
    sum->animation= this->animation + dynamic_cast<Skill*>(s)->animation;
    sum->manaCost = this->manaCost + dynamic_cast<Skill*>(s)->manaCost;

    Damage* ret = sum;

    return ret;

}

double Skill::getMaxAnim(Skill* s) const{
    if(this->animation > s->animation)
        return this->animation;
    else
        return s->animation;
}


unsigned int Skill::getLevel() const{
    return level;
}

File: SAUCE/Java/Skill.java
public class Skill extends Damage {
    private boolean ready;
    private string Sname;
    private int manaCost;
    private int coolDown;
    private int range;
    private int level;
    private double projectileSpeed;
    private double animation;
    public Skill(double v,string s,int m,double an,int cd,int r,double ps,int lv){
        super(v);
        re=true;
        Sname=s;
        manaCost=m;
        coolDown=cd;
        range=r;
        level=lv;
        projectileSpeed=ps;
        animation=an;
     }

    public boolean IsCastable(double time,int distance){
        if(time<animation+(distance/projectileSpeed) || distance>range || (level<=0 || level>4))
            return false;
        else
            return true;
        }

    public boolean IsCastable(double time,int distance,int mana){
        if(time<animation+(distance/projectileSpeed) || distance>range || (level<=0 || level>4)||mana<manaCost)
            return false;
        else
            return true;
        }

    public double DPS(int distance){
        return DamageByTime(1,distance);
        }

    public double DamageByTime(double time,int distance){
        return HitByTime(time,distance)*value;
        }

    public int HitByTime(double time,int distance){
        double aux = time;
        unsigned int tiri = 0;
        while(IsCastable(aux,distance)){
            tiri++;
            aux=aux-(animation+cooldown+(distance/projectileSpeed));
            }
        return tiri;
        }

    public double totalTime(int distance){
        return (distance/projectileSpeed)+animation;
        }

    public double getAnimation(){
        return animation;
        }

    public boolean getReady(){
        return ready;
        }

    public int getCooldown(){
        return coolDown;
        }
    public int getManaCost(){
        return manaCost;
        }

    public string getName(){
        return Sname;
        }

    void setReady(bool b){
        ready=b;
        }

    public boolean equals(Object o){
        return(o instanceof Skill)&&(getAnim()==((Skill)o).getAnim())&&(getRange()==((Skill)o).getRange())&&(super.getValue()==d.getValue())&&(getManaCost()==((Skill)o).getManaCost())&&(getManaCost()==((Skill)o).getManaCost());
        }

    public double sum(Skill s){
        return getValue()+s.getValue();
        }

    public double dif(Skill s){
        return getValue()-s.getValue();
        }
}
